### PR TITLE
Refactoring of the Objective-C integration

### PR DESCRIPTION
**Pull request by jacob-carlborg**
I'm not sure if there's a better place for the functions in `objc_glue.c`.

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
@yebblies Refactoring of the Objective-C integration

**Comment by WalterBright** [comment]
Can you give a brief description of the purpose of the refactoring? My cursory look at it is it appears to move ObjC logic out of objc modules and dispersing it into the rest of the front end. Why?

**Comment by jacob-carlborg** [comment]
* Currently the Objective-C integration is the only feature that is organized in this, slightly weird, way. With this refactoring it brings the feature up to the same level as the rest of the features

* It doesn't work with cross-compiling which is important for LDC and might become important later for DMD as well. Currently all the Objective-C logic is removed at compile time if the target doesn't support Objective-C. The decision if Objective-C is supported or not need to occur at runtime and not compile time for cross-compiling to work

* This all the functions with the weird names that was a result of being force to moved out logic that should have been inline in the rest of the code

* It simplifies the design because `objc.d`, `objc_stubs.d` and `objc_glue_stubs.c` are all removed

I talked with Daniel and Iain during dconf about this and they both think that this is the right decision. I'm pretty sure that David agrees as well.

**Comment by jacob-carlborg** [comment]
Hmm, seems I need to give this refactoring some more thought.

**Comment by WalterBright** [comment]
> the only feature that is organized in this, slightly weird, way.

No, it's not. Header generation, json generation, ddoc, array ops, etc., are all factored out into separate modules. I would like to increase this sort of encapsulation, not decrease it. The main modules are already far too large and convoluted. Distributing the objc support among them will make things more convoluted and harder to understand.

Your other goals with this are good, but I do not believe they require distributing their contents around the rest of the code to achieve them.

**Comment by yebblies** [comment]
Those other features are not adding new types, semantic processing, etc.  We do not have a nice system for factoring out extern language support in dmd, and rather than making it clearer this layout just makes a mess.  An entire function for two lines of code?  It's a joke, and it doesn't work.

Either this is a first-class feature that deserves being integrated with the rest of dmd, or we should delete it.

**Comment by ibuclaw** [comment]
Objective-C implementation is pretty dumb, I couldn't find a way to integrate it into GDC without https://github.com/dlang/dmd/pull/4813

**Comment by ibuclaw** [comment]
Everything should have been put into an `Objc` namespace from the start.  extern C functions are not a form of encapsulation.

**Comment by WalterBright** [comment]
I am interested in proper encapsulation in order to reduce bugs and complexity. I am not interested in status symbols of where code is located. If objc is support is lacking (and I don't know if it is or is not) we can address that.

> An entire function for two lines of code? It's a joke, and it doesn't work.

We have lots of functions in dmd that are a couple lines. They serve encapsulation purposes. As I recall you asked for a one-liner in root/outbuffer yesterday (a `length()` function). Besides, func.d is acquiring a number of functions with many more than 2 lines of code in them.

The front end, in my not-so-humble opinion, has major problems with lack of encapsulation. The lack of encapsulation is not a justification for even less encapsulation.

**Comment by jacob-carlborg** [comment]
> If objc is support is lacking (and I don't know if it is or is not) we can address that

That's what I intend to work on. After the huge 12k+ lines of code PR which was the initial one, we decided to split it up in more manageable pieces. The rest of the Objective-C support has not been integrated yet. I talked with Daniel at dconf and we agreed that we should do this refactoring first (assuming this is what he had in mind).

**Comment by jacob-carlborg** [comment]
> "Objective-C support" is an useful unit of encapsulation

The full integration will touch many (most) parts of the compiler.

File: src/aggregate.h [hunk 1]
```diff
@@ -279,11 +278,11 @@ class ClassDeclaration : public AggregateDeclaration
     TypeInfoClassDeclaration *vclassinfo;       // the ClassInfo object for this ClassDeclaration
     bool com;                           // true if this is a COM class (meaning it derives from IUnknown)
     bool cpp;                           // true if this is a C++ interface
+    bool objc;                          // true if this is an Objective-C class/interface
```
Comment (ibuclaw) [inline src/aggregate.h]: Classes can only be `com`, `cpp`, or `objc`, but never two at once, right?  These bool flags could be squashed into an enum later, perhaps.

Comment (jacob-carlborg) [inline src/aggregate.h]: Yes, as far as I know. If it's `objc` it cannot be anything else at least. I guess a new PR would be a better fit for that.

**Comment by yebblies** [comment]
> I am interested in proper encapsulation in order to reduce bugs and complexity. I am not interested in status symbols of where code is located. If objc is support is lacking (and I don't know if it is or is not) we can address that.

My point is that unless we're planning to delete it (and therefore want it nicely isolated and easy to delete) then we should put the parts where they logically make sense - next to the handling of other extern types in dmd.

> We have lots of functions in dmd that are a couple lines. They serve encapsulation purposes.

These functions do not serve encapsulation purposes.  They are simply manually outlined chunks of code, that only have one caller.  If your concern is that it might not be immediately obvious what they're for, then adding comments would be a less disruptive solution.

> As I recall you asked for a one-liner in root/outbuffer yesterday (a length() function).

Because it give a name to a common pattern... there is a difference here.

> The front end, in my not-so-humble opinion, has major problems with lack of encapsulation. The lack of encapsulation is not a justification for even less encapsulation.

I would agree with you if I thought that outlining objective-C support into another file improved encapsulation.  Instead it just hides the code from people editing the related functions, and adds more false dependencies between modules.

Just look at how the line count changes...

**Comment by jacob-carlborg** [comment]
BTW, I added documentation for most of the remaining functions.

Any reason why the Windows builds are still pending? It looks like it already has built twice on the other platforms.

File: src/aggregate.h (hunk 1 above)
Comment (yebblies) [inline src/aggregate.h]: IIRC there is at least one other way to mark a class as c++, it could do with some cleaning up.

**Comment by jacob-carlborg** [comment]
I assume the Windows failures are not related: `core.exception.AssertError@std\datetime.d(28135): TZName which is missing: Easter Island Standard Time`

**Comment by Hackerpilot** [comment]
@jacob-carlborg: https://github.com/dlang/phobos/pull/4317

**Comment by jacob-carlborg** [comment]
@Hackerpilot thanks. Can we rerun the Windows tests?

**Comment by jacob-carlborg** [comment]
Awesome, all tests are passing now.

File: src/declaration.h [hunk 2]
```diff
@@ -885,4 +885,24 @@ class DeleteDeclaration : public FuncDeclaration
     void accept(Visitor *v) { v->visit(this); }
 };
 
+struct ObjcSelector
```
Comment (WalterBright) [inline src/declaration.h]: This should go in objc.d, there's no reason to make the fields and methods available throughout the compiler.

File: src/declaration.h [hunk 3]
```diff
@@ -885,4 +885,24 @@ class DeleteDeclaration : public FuncDeclaration
...
+{
+    static StringTable stringtable;
```
Comment (WalterBright) [inline src/declaration.h]: None of these fields and members appear to be documented.

File: src/e2ir.c [hunk 4]
```diff
@@ -243,8 +242,21 @@ elem *callfunc(Loc loc,
         }
     }
 
-    objc_callfunc_setupMethodSelector(tret, fd, t, ehidden, &esel);
-    objc_callfunc_setupEp(esel, &ep, reverse);
+    if (global.params.hasObjectiveC)
```
Comment (WalterBright) [inline src/e2ir.c]: I don't see how moving the code here helps.

File: src/e2ir.c [hunk 5]
```diff
@@ -305,9 +317,12 @@ elem *callfunc(Loc loc,
         }
         Symbol *sfunc = toSymbol(fd);
 
-        if (esel)
+        if (global.params.hasObjectiveC && esel)
```
Comment (WalterBright) [inline src/e2ir.c]: I don't see how this is an improvement.

File: src/func.d [hunk 6]
```diff
@@ -1335,14 +1333,106 @@ public:
...
 
+    /**
+     * Sets the Objective-C selector of the receiver, if present.
+     *
+     * The Objective-C selector is extracted from the `@selector` UDA.
+     *
+     * Params:
+     *  sc = the current scope
+     */
+    private void setSelector(Scope* sc)
```
Comment (WalterBright) [inline src/func.d]: Private objc functions should go in objc.d.

File: src/func.d [hunk 7]
```diff
@@ -5766,3 +5856,156 @@ public:
         v.visit(this);
     }
 }
+
+/***********************************************************
+ * This class represents an Objective-C selector.
```
Comment (WalterBright) [inline src/func.d]: should go in objc.d

File: src/objc_glue.c [hunk 8]
```diff
@@ -107,6 +142,29 @@ void objc_initSymbols()
         objc_segList[s] = 0;
 }
 
+/**
+ * Returns `true` if the compiler have seen any Objective-C symbols.
```
Comment (WalterBright) [inline src/objc_glue.c]: s/have/has/

File: src/objc_glue.c [hunk 9]
```diff
@@ -107,6 +142,29 @@ void objc_initSymbols()
...
+ * This is used to decides if Objective-C module info should be generated.
```
Comment (WalterBright) [inline src/objc_glue.c]: s/decides/decide/

**Comment by WalterBright** [comment]
I object to moving the code into func.d. I don't see how that improves anything. It reduces encapsulation by making the objc structs and members visible everywhere. I don't see how it makes anything easier to understand.

File: src/declaration.h (hunk 3 above)
Comment (jacob-carlborg) [inline src/declaration.h]: It's documented in `func.d`. Do I need to duplicate the documentation?

File: src/objc_glue.c (hunk 8 above)
Comment (jacob-carlborg) [inline src/objc_glue.c]: Fixed.

File: src/objc_glue.c (hunk 9 above)
Comment (jacob-carlborg) [inline src/objc_glue.c]: Fixed.

**Comment by jacob-carlborg** [comment]
Could we please come to some form of agreement here. I would like to continue with the Objective-C integration but we need to decide how to structure the code.

**Comment by WalterBright** [comment]
> It simplifies the design because objc.d, objc_stubs.d and objc_glue_stubs.c are all removed

Removing files has no correlation with simplifying the design. But I see no reason why the ObjC logic cannot all be contained in objc.d, objc.h, and objc_glue.c. Then, one only has 3 files to look at to follow the ObjC design and figure out how it works.

**Comment by ibuclaw** [comment]
> Could we please come to some form of agreement here?

I would say that all data structures and functions prefixed with `objc_ClassDeclaration` and `objc_FuncDeclaration` should be first class citizens of their respective classes.  Having an objc param in `globals` is not strictly necessary as we can put it elsewhere, such as in `Objc::isSupported`.

Other routines that hide the internals implementation should be in an `Objc` interface. Think how Target or Port works.

**Comment by yebblies** [comment]
> Having an objc param in `globals` is not strictly necessary as we can put it elsewhere, such as in `Objc::isSupported`.

I'd rather have global state in globals where possible.

**Comment by jacob-carlborg** [comment]
Yet another PR that is stalled. @WalterBright you seem to be the only one agains this change.

**Comment by WalterBright** [comment]
Yes, I am against it, for reasons stated.

**Comment by jacob-carlborg** [comment]
@WalterBright @yebblies what if we do a compromise. If I inline calls like `objc_ClassDeclaration_semantic_PASSinit_LINKobjc` and `objc_callfunc_setupMethodSelector` but keep code like `struct ObjcSelector` and `validateSelector` in the `objc.d/c` file?

**Comment by jacob-carlborg** [comment]
> I don't understand what the proposed compromise does.

@WalterBright  I have not changed any code yet in the PR. Currently the definition of `objc_ClassDeclaration_semantic_PASSinit_LINKobjc` looks like this [1] and is used here [2]. My proposed compromise is to move that function inline at the call site to look like this (in dclass.d):

```d
if (sc.linkage == LINKobjc)
{
    if (global.params.hasObjectiveC)
        objc = true;
    else
        error("Objective-C classes are not supported for this target");
}
```

Other symbols like `ObjcSelector` would be moved back to `objc.d`. If you think it's ok that the `objc_ClassDeclaration_semantic_PASSinit_LINKobjc` function is inline at the call site I can update the PR to show how it would look like with the rest of the symbols.

[1] https://github.com/dlang/dmd/blob/master/src/objc.d#L143
[2] https://github.com/dlang/dmd/blob/master/src/dclass.d#L500-L501

**Comment by WalterBright** [comment]
Thanks for the explanation. I honestly don't see the point of making this change. How does it improve things?

**Comment by jacob-carlborg** [comment]
> Thanks for the explanation. I honestly don't see the point of making this change. How does it improve things?

@WalterBright it would remove the awkward functions like `objc_ClassDeclaration_semantic_PASSinit_LINKobjc`. They're awkward because they should really not be separate functions. They're only called once and doesn't provide any form of abstraction.

**Comment by wilzbach** [comment]
> JC: Could we please come to some form of agreement here. I would like to continue with the Objective-C integration but we need to decide how to structure the code.
[...]
> JC: Yet another PR that is stalled. @WalterBright you seem to be the only one against this change.
> WB: Yes, I am against it, for reasons stated.

@jacob-carlborg did you ever come to an agreement or is this still stalled?

**Comment by ibuclaw** [comment]
This is still stalled as far as I know.  Which is a shame I think it's in the interest of everyone to fix this interface.

**Comment by jacob-carlborg** [comment]
@wilzbach it's still stalled. Therefore the rest of Objective-C integration is stalled.

**Pull request closed without merging**